File: leetcode/763-partition-labels/main.py
from typing import List, Dict

class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        ends: Dict[str, int] = {}
        for i in range(len(s)):
            ends[s[i]] = i

        partitions = []
        prev, current = 0, 0
        for i in range(len(s)):
            current = max(current, ends[s[i]])
            if i == current:
                partitions.append(current-prev+1)
                prev = current+1

        return partitions


solution = Solution()
s = "ababcbacadefegdehijhklij"
print(solution.partitionLabels(s))

File: leetcode/763-partition-labels/main.go
package main

import (
	"fmt"
)

func partitionLabels(s string) []int {
	ends := make(map[byte]int)
	for i := range s {
		ends[s[i]] = i
	}
	partitions := make([]int, 0)
	var prev, current int
	for i := range s {
		end := ends[s[i]]
		if end > current {
			current = end
		}
		if i == current {
			partitions = append(partitions, current-prev+1)
			prev = current + 1
			continue
		}
	}
	return partitions
}

func main() {
	s := "ababcbacadefegdehijhklij"
	fmt.Println(partitionLabels(s))
}
